examples/keras_recipes/memory_efficient_embeddings.py

```diff
@@ -0,0 +1,419 @@
+"""
+
+import os
+import math
+from zipfile import ZipFile
+from urllib.request import urlretrieve
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.keras.layers.experimental.preprocessing import StringLookup
+import matplotlib.pyplot as plt
+
+"""
+## Prepare the data
+
+## Download and process data
+"""
+
+urlretrieve("http://files.grouplens.org/datasets/movielens/ml-1m.zip", "movielens.zip")
+ZipFile("movielens.zip", "r").extractall()
+
+ratings_data = pd.read_csv(
+    "ml-1m/ratings.dat",
+    sep="::",
+    names=["user_id", "movie_id", "rating", "unix_timestamp"],
+)
+
+ratings_data["movie_id"] = ratings_data["movie_id"].apply(lambda x: f"movie_{x}")
+ratings_data["user_id"] = ratings_data["user_id"].apply(lambda x: f"user_{x}")
+ratings_data["rating"] = ratings_data["rating"].apply(lambda x: float(x))
+del ratings_data["unix_timestamp"]
+
+print(f"Number of users: {len(ratings_data.user_id.unique())}")
+print(f"Number of movies: {len(ratings_data.movie_id.unique())}")
+print(f"Number of ratings: {len(ratings_data.index)}")
+
+"""
+## Create train and eval data splits
+"""
+
+random_selection = np.random.rand(len(ratings_data.index)) <= 0.85
+train_data = ratings_data[random_selection]
+eval_data = ratings_data[~random_selection]
+
+train_data.to_csv("train_data.csv", index=False, sep="|", header=False)
+eval_data.to_csv("eval_data.csv", index=False, sep="|", header=False)
+print(f"Train data split: {len(train_data.index)}")
+print(f"Eval data split: {len(eval_data.index)}")
+print("Train and eval data files are saved.")
+
+"""
+## Define dataset metadata and hyperparameters
+"""
+
+csv_header = list(ratings_data.columns)
+user_vocabulary = list(ratings_data.user_id.unique())
+movie_vocabulary = list(ratings_data.movie_id.unique())
+target_feature_name = "rating"
+learning_rate = 0.001
+batch_size = 128
+num_epochs = 3
+base_embedding_dim = 64
+
+"""
+## Train and evaluate the model
+"""
+
+
+def get_dataset_from_csv(csv_file_path, batch_size=128, shuffle=True):
+    return tf.data.experimental.make_csv_dataset(
+        csv_file_path,
+        batch_size=batch_size,
+        column_names=csv_header,
+        label_name=target_feature_name,
+        num_epochs=1,
+        header=False,
+        field_delim="|",
+        shuffle=shuffle,
+    )
+
+
+def run_experiment(model):
+    # Compile the model.
+    model.compile(
+        optimizer=keras.optimizers.Adam(learning_rate),
+        loss=tf.keras.losses.MeanSquaredError(),
+        metrics=[keras.metrics.MeanAbsoluteError(name="mae")],
+    )
+    # Read the training data.
+    train_dataset = get_dataset_from_csv("train_data.csv", batch_size)
+    # Read the test data.
+    eval_dataset = get_dataset_from_csv("eval_data.csv", batch_size, shuffle=False)
+    # Fit the model with the training data.
+    history = model.fit(train_dataset, epochs=num_epochs, validation_data=eval_dataset)
```

Comment:
You don't know in advance which epoch will perform best, so use a ModelCheckpoint and EarlyStopping callback (with a large number of epochs) to stop training after the model starts overfitting, then recover the best model